tools/test-runner/src/buck_test.rs

```diff
@@ -0,0 +1,252 @@
+            return TestResult {
+                name: test.name,
+                attempts,
+                passed: true,
+                duration,
+                stdout: out,
+                stderr: err,
+                contacts: test.contacts,
+            };
+        } else {
+            if attempts >= 1 + retries {
+                let mut out = String::new();
+                child.stdout.unwrap().read_to_string(&mut out).unwrap();
+                let mut err = String::new();
+                child.stderr.unwrap().read_to_string(&mut err).unwrap();
+                return TestResult {
+                    name: test.name,
+                    attempts,
+                    passed: false,
+                    duration,
+                    stdout: out,
+                    stderr: err,
+                    contacts: test.contacts,
+                };
+            }
+        }
+    }
+}
+
+// a.k.a., the defaults
+pub mod shell {
+    use super::{make_command, Test, TestKind, TestSpec};
+    use std::process::Child;
+
+    /// Builds a singleton test from a given spec.
+    pub fn validate(spec: TestSpec) -> Vec<Test> {
+        let command = make_command(
+            spec.command[0].clone(),
+            spec.command[1..].to_vec(),
+            spec.env,
+            spec.cwd,
+        );
+        let test = Test {
+            command,
+            name: spec.target,
+            labels: spec.labels,
+            contacts: spec.contacts,
+            kind: TestKind::Shell,
+        };
+        return vec![test]; // i.e; "just run it"
+    }
+
+    /// Validates the test based solely on its return value.
+    pub fn evaluate(result: &mut Child) -> bool {
+        result.wait().unwrap().success() // shouldn't block, it already ran
+    }
+}
+
+/// Refer to https://buck.build/files-and-dirs/buckconfig.html#test.external_runner
+///
+/// A JSON spec needs at least these fields, any others will be ignored.
+#[derive(Clone, Debug, Deserialize)]
+pub struct TestSpec {
+    /// The buck target URI of the test rule.
+    pub target: String,
+
+    /// The type of the test.
+    #[serde(rename(deserialize = "type"), default = "default_kind")]
+    kind: TestKind,
+
+    /// Command line that should be used to run the test.
+    pub command: Vec<String>,
+
+    /// Environment variables to be defined when running the test.
+    pub env: HashMap<String, String>,
+
+    /// Labels that are defined on the test rule.
+    pub labels: HashSet<String>,
+
+    /// Contacts that are defined on the test rule.
+    pub contacts: HashSet<String>,
+
+    /// Working directory the test should be run from.
+    pub cwd: Option<PathBuf>,
+
+    /// Absolute paths to any files required by this test target.
+    required_paths: Option<Vec<PathBuf>>,
+}
+fn default_kind() -> TestKind {
+    TestKind::Shell
+}
+
+/// Reads test specs from a buck test run description at the given path.
+pub fn read(path: &PathBuf) -> Result<Vec<TestSpec>> {
```

Comment:
The rust convention here would be to use a type parameter on `path` to accept all different kinds of `Path` objects in a method like this:

```
pub fn read<P: AsRef<Path>>(path: P) -> Result<Vec<TestSpec>> {
```

and then calling `path.as_ref()` to get a `&Path` object. Then your function can accept anything that implements `AsRef<Path>` (`str`, `String`, `OsStr`, `OsString`, `Path` and `PathBuf` in the stdlib to name the most common)

Reply:
That's quite handy, thanks for the heads up!